### ISSUE TITLE
Some questions about Paper‘s

### ISSUE THREAD

**Issue by qj1204** [WindVChen/DiffAttack#19]
Hello, I am currently reproducing your paper. Regarding Figure 4, I have some questions that I would like to ask you:
1. Is the dataset Imagene-compatible?
2. Besides DiffAttack, which specific surrogate model is attacked by other methods to obtain adversarial samples (which one of Res-50, VGG-19, Mov-v2, Inc-v3, ConvNeXt, and Swin-B)?
3. Are other methods using third-party adversarial attack libraries or running only source code?
![image](https://github.com/WindVChen/DiffAttack/assets/49404258/4622d161-6f66-4a10-81a8-ccd2c6146d5c)

There is another question about table 1:
The gray background represents a white box attack, but DiffAttack uses the adversarial samples generated by the diffusion model to attack the target model, which should be a black box attack. Why is DiffAttack also set to gray in Table 1?
![image](https://github.com/WindVChen/DiffAttack/assets/49404258/de70f4d7-dab5-4078-a076-d485ab8464bf)

**Comment by WindVChen** [comment]
Hi @qj1204 ,

1. Yes. Visualizations in Fig. 4, Fig. 8, and Fig. 9 are from the ImageNet-Compatible dataset.
2. All are crafted on Inc-v3, including DiffAttack.
3. Please refer to this issue https://github.com/WindVChen/DiffAttack/issues/3#issuecomment-1565459090.
4. There may be some misunderstandings. Although the adversarial samples of DiffAttack are generated by the diffusion model, the optimization of them received guidance from the surrogate classifier (refer to Eq. 3 in the paper). Thus, according to the definition of white-box attacks, when the surrogate model is the same as the target model, the attack is a white-box attack.

Hope these help.

**Comment by qj1204** [comment]
Thank you for your reply. Now I have a clearer understanding of the DiffAttack's process.

There is one more question. The image is part of the SSA code. Yesterday, I tried to run it, but I don't know if line 108 needs to be deleted when running DI, TI, MI. May I ask which lines you specifically deleted or released at that time?
![image](https://github.com/WindVChen/DiffAttack/assets/49404258/09a60d1b-76e6-48e0-be0e-ca172804079d)

**Comment by qj1204** [comment]
And Why is there a huge difference between my results and those in your paper? The FID value also has a slight difference. The surrogate model I am using is Inc-v3, and the method is S²I-FGSM.
![image](https://github.com/WindVChen/DiffAttack/assets/49404258/8ecef51b-ee1a-409d-b81a-8c97f4ac961c)
![image](https://github.com/WindVChen/DiffAttack/assets/49404258/422b722b-5831-4471-9db4-bb2fe42116fe)
![image](https://github.com/WindVChen/DiffAttack/assets/49404258/49bfdf62-a9ba-43d2-9f6d-096879fe79cf)![image](https://github.com/WindVChen/DiffAttack/assets/49404258/22f2e4c3-338a-454a-af0d-b8abc346ef8d)

**Comment by WindVChen** [comment]
Hi @qj1204 ,

I've written a brief script that you can use to replace the original lines 87-116:

```python
for i in range(num_iter):
    noise = 0

    x_idct = V(x.cuda(), requires_grad=True)
    
    # DI-FGSM https://arxiv.org/abs/1803.06978
    # output_v3 = model(DI(x_idct))
    output_v3 = model(x_idct)

    loss = F.cross_entropy(output_v3, gt)
    loss.backward()
    noise += x_idct.grad.data

    # TI-FGSM https://arxiv.org/pdf/1904.02884.pdf
    # noise = F.conv2d(noise, T_kernel, bias=None, stride=1, padding=(3, 3), groups=3)

    # MI-FGSM https://arxiv.org/pdf/1710.06081.pdf
    # noise = noise / torch.abs(noise).mean([1, 2, 3], keepdim=True)
    # noise = momentum * grad + noise
    # grad = noise

    x = x + alpha * torch.sign(noise)
    x = clip_by_tensor(x, min, max)

return x.detach()
```

The script above essentially implements an iterative FGSM (with value clipping). By uncommenting the lines under the respective DI-FGSM, TI-FGSM, and MI-FGSM sections, the code will be modified accordingly. Note that for DI-FGSM, you need to comment out the line `output_v3 = model(x_idct)`.

I hope this helps!

By the way, I noticed that you previously also attached a figure of test results. However, I don't see it now. Has that problem been resolved?

**Comment by qj1204** [comment]
That picture shows that there is a significant difference between my test results and the results in the paper. Later, I discovered that the order of the images did not match the label, and now I have resolved this issue.
However, the current results are slightly higher than those in the paper.

The surrogate model is Inc-v3, and the attack method is S²I-FGSM.
My result:
![image](https://github.com/WindVChen/DiffAttack/assets/49404258/274738c3-e2cd-463f-b477-2d45a437aa78)
Paper's result:
![image](https://github.com/WindVChen/DiffAttack/assets/49404258/44db2981-8540-43d9-8baf-8b937b59ca12)

**Comment by qj1204** [comment]
May I ask how did you set up the model when reproducing S²I-FGSM?

The ignored part is the source code of S²I-FGSM. If I don't modify it to line 150, the resolution cannot be changed to 224. Otherwise, an error will be reported. After adding line 150 and changing the resolution to 224, I obtained the above result.
![image](https://github.com/WindVChen/DiffAttack/assets/49404258/5c82f022-ba04-4e3b-8980-2d8a7eb5b15a)

**Comment by WindVChen** [comment]
Ah, you may need to include the normalization with L150.

**Comment by qj1204** [comment]
Sorry, I don't quite understand what you meant. Can you explain it in more detail? 
Please forgive me for bothering you all this time.

**Comment by WindVChen** [comment]
You may have a try with this (incorporate the normalization operation into the line 150):
```python
model = torch.nn.Sequential(Normalize(opt.mean, opt.std),
                  models.inception_v3(pretrained=True).eval().cuda())
```

**Comment by qj1204** [comment]
Thank you for your suggestion. The results I have reproduced now are almost identical to yours.
By the way, how do you run [HGD](https://github.com/lfz/Guided-Denoise) code? I have adversarial samples generated by various attack methods. Should I directly run 'python main. py -- exp sample'?

**Comment by WindVChen** [comment]
Hi,

For HGD, you should run `defense.py`. You can find more information [here](https://github.com/lfz/Guided-Denoise/tree/master/nips_deploy).

Regarding DiffPure, you can refer to DiffPure's source code for details. In the adversarial purification evaluation, they first generate adversarial samples and then pass these samples through the diffusion denoising process. To integrate this with other attacks, you can skip the initial adversarial sample generation step and instead use adversarial samples generated by other attack methods, such as our DiffAttack, and then send these to their diffusion denoising process.
